File: libs/power-log-parser/src/lib/cards/runic-adornment.ts
import { BlockType, GameTag, Zone } from '@firestone-hs/reference-data';
import { Action } from '../models/action';
import { TagChange } from '../models/tag';
import { Node, NodeType } from '../models/node';
import type { GameState } from '../state/game-state';
import type { StateFacade } from '../state/state-facade';

export class RunicAdornment {
	static PredictCardId(
		gameState: GameState,
		creatorCardId: string,
		creatorEntityId: number,
		node: Node,
		stateFacade: StateFacade | null,
	): string | null {
		if (node.Parent?.Type !== NodeType.Action) {
			return null;
		}

		const act = node.Parent.Parent?.Object as Action;
		if (act.Type !== (BlockType.PLAY as number)) {
			return null;
		}

		const triggerAction = act.Data
			.filter((d): d is Action => d instanceof Action)
			.find((a) => a.Type === (BlockType.POWER as number) && a.Entity === act.Entity);
		if (!triggerAction) {
			return null;
		}

		const tagChange = triggerAction.Data
			.filter((d): d is TagChange => d instanceof TagChange)
			.find(
				(d) =>
					d.Name === (GameTag.ZONE as number) && d.Value === (Zone.HAND as number),
			);
		if (!tagChange) {
			return null;
		}

		const entity = gameState.CurrentEntities.get(tagChange.Entity);
		return entity?.CardId ?? null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Cards/RunicAdornment.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using HearthstoneReplays.Parser.ReplayData.Entities;

namespace HearthstoneReplays.Events.Cards
{
    internal class RunicAdornment
    {
        public static string PredctCardId(GameState gameState, string creatorCardId, int creatorEntityId, Node node, StateFacade stateFacade)
        {
            if (node.Parent?.Type != typeof(Action))
            {
                return null;
            }
            
            var act = node.Parent.Parent.Object as Action;
            if (act.Type != (int)BlockType.PLAY)
            {
                return null;
            }

            var triggerAction = act.Data
                .Where(d => d is Action)
                .Select(d => d as Action)
                .Where(a => a.Type == (int)BlockType.POWER && a.Entity == act.Entity)
                .FirstOrDefault();
            if (triggerAction == null)
            {
                return null;
            }

            var tagChange = triggerAction.Data
                .Where(d => d is TagChange)
                .Select(d => d as TagChange)
                .Where(d => d.Name == (int)GameTag.ZONE && d.Value == (int)Zone.HAND)
                .FirstOrDefault();
            if (tagChange == null)
            {
                return null;
            }

            var entity = gameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);
            return entity?.CardId;
        }
    }
}
